services/queue/src/ec2regionresolver.js

```diff
@@ -9,19 +9,59 @@ const AWS_IP_RANGES_URL = 'https://ip-ranges.amazonaws.com/ip-ranges.json';
 /** Resolve EC2 region from request using AWS ip-ranges */
 class EC2RegionResolver {
   /** Construct EC2RegionResolver given a list of regions we care about */
-  constructor(regions) {
+  constructor(regions, monitor) {
     assert(regions instanceof Array, 'regions must be an array');
     this.regions = regions;
+    this.monitor = monitor;
     this.ipRanges = [];
   }
 
-  async loadIpRanges() {
-    // Get IP ranges from AWS with really stupid retry logic
-    let {body} = await request.get(AWS_IP_RANGES_URL).catch(() => {
-      return request.get(AWS_IP_RANGES_URL).catch(() => {
-        return request.get(AWS_IP_RANGES_URL);
-      });
+  /**
+   * Start running the resolver.  This will return immediately while loading the
+   * list of IP ranges.  Until that load is complete, the resolver will return
+   * null (meaning not an EC2 region) for all IPs.
+   */
+  start() {
+    let delay = 500;
+    this._running = true;
+
+    this._loadPromise = new Promise(resolve => {
+      const tryLoad = async () => {
+        if (!this._running) {
+          return resolve();
+        }
+
+        try {
+          await this._loadIpRanges();
+          return resolve();
+        } catch (err) {
+          this.monitor.warning(`Failed to download AWS IP ranges (retrying): ${err}`);
+          setTimeout(tryLoad, delay);
+          delay *= 2;
+        }
+      };
+      tryLoad();
     });
```

Comment:
I think we should throw an error after some max time. It can be very long if we want, perhaps 10 minutes even? Still would be nice to crash eventually since potentially this is a very costly issue.

Another option would be to change the warning to an error after 10 attempts or something?

Reply:
I'm not sure what we would do armed with that information.  Crashing isn't great as it just expands the failure radius -- from a few too-expensive artifact URLs to a total failure of the queue service.

One option would be to cache a copy of the region information in the source, and default to that until we can successfully download the most up-to-date region information.  I can try that out..

Reply:
If it is really just a handful of too-expensive calls maybe we should just remove this feature for now?

Reply:
The feature is useful when cloudfront isn't down.. I just mean that during the time cloudfront is down, we would rather make a few too-expensive calls than crash.